### ISSUE TITLE
Provide a better reference for what units are supported.

### ISSUE THREAD

**Issue by ChrisBarker-NOAA** [cf-convention/cf-conventions#562]
# Provide a better reference for what units are supported.

## Moderator
TBD

### Moderator Status Review [last updated: YYYY-MM-DD]

Brief comment on current status, update periodically

## Requirement Summary

Users of CF need to know what unit strings  they can use for their data. While the standard names provide canonical units, others are allowed, not everything has a standard name. 

Consumers of CF data files also need to know what unit strings they might encounter for given data.

So there should be an easy way for users -- both producers and consumers, to quickly know what the options are for a particular physical quantity, e.g -- this data variable is a velocity, I can use (and may encounter) `m/s` or `meters per seconds` or `meter/s` or [a large number of combinations!] 

CF passes this responsibility on to the UDUNITS package -- that's fine and complete, but the documentation is lacking. 

## Technical Proposal Summary

The CF spec should include, or reference, a clear and complete definition of all the units and spellings that are allowed.

It's possible that such a document exists, in which case, CF could simply point to it.

Otherwise, a new document / page / seciton should be created.

## Benefits

Everyone that is producing CF compliant files, and even more so folks that are writing code to consume CF compliant files.

## Status Quo
Currently in CF, there is this text:

> The value of the units attribute is a string that can be recognized by the UDUNITS package [[UDUNITS]](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.11/cf-conventions.html#UDUNITS), 
...
>  Note that case is significant in the units strings. Note also that CF depends on UDUNITS only for the definition of legal units strings. CF does not assume or require that the UDUNITS software will be used for units conversion.

There is a link to the UDUNITS page:

https://www.unidata.ucar.edu/software/udunits/

Technically, this is complete, but in practice, the UDUNITS page isn't that helpful for letting people know what they can use. All the definitions are in XML files, but that's not exactly human-readable. And I don't see anywhere there rules for pluralization, how to combine units, etc, etc.

So the only way really to know is to use UDUNITS itself to test what you may want to use, which is a bit in conflict with the statement that "CF does not assume or require that the UDUNITS software will be used for units conversion."

The challenge here is that UDUNITS accepts a LOT of spellings for units, e.g.  "m", "meter", "meters" -- "m/s", "meters per second" , "m.s-1".

It is very hard to know what the all options are.

And there are no recommendations for a preferred spelling of units,  beyond the canonical units used in standard names.

# Associated pull request

TBD

## Detailed Proposal
[Incomplete as of now]

### Part 1: Provide a clear and complete description of the rules governing unit strings

This can be a link to another doc in UDUNITS (or elsewhere) if it exists -- in which case, this is easy, but I haven't found one.

Or we'll need to write a new doc that lays it out -- ideally with the help of the UDUNITS folks.

### part 2: Provide recommendations for spelling of units.

Postel's law: "be conservative in what you send, be liberal in what you accept".

In this context that means UDUNITS is doing the right thing in accepting a wide variety of spellings for units.

But CF is not being "conservative in what you send"

In order to do that, CF would define (too late for that) or recommend (worth it?) a subset of the UDUNITS accepted spellings.

That is, should someone use:

"m.s-1" or "meter/second" or "meters per second" or ???? 

Perhaps that ship has sailed long, long ago, but is it a bad idea to start nailing it down more now? maybe.

**Comment by larsbarring** [comment]
Further thought: is this better placed as discussion?

**Comment by roy-lowry** [comment]
This issue was addressed (or circumvented) in the SeaDataNet 'layer' over CF by adding a SeaDataNet namespace parameter attribute that holds the URI for the unit of measure in a controlled vocabulary (https://vocab.nerc.ac.uk/search_nvs/P06/).

**Comment by larsbarring** [comment]
Roy, thanks for reviving this issue which I had totally forgotten. Since my previous comment I have spent a bit of time digging into the UDUNITS2 xml units database, and also scratched on the surface of the C code base. One particular aim was to update the units -- and not least their description -- with the 2019 redefinition of the SI units. Here is a link to the github pages that resulted from this activity: 
https://udusi.github.io/UDUSI/ 
It is still early days, but I hope reasonably polished (hence this comment :-).

**Comment by JonathanGregory** [comment]
UDUSI is a remarkable and impressive piece of work! Congratulations, @larsbarring!

**Comment by efisher008** [comment]
Is this conversation now concluded? Is there anything the contributors would like to bring up/highlight at the 2025 CF workshop, perhaps?

**Comment by JonathanGregory** [comment]
We had a discussion at the CF committee meeting on 15th April about whether the CF conventions document should explicitly state the UDUNITS syntax, to make it more accessible (perhaps in an appendix), and whether CF should provide its own units database (perhaps related to Lars's UDUSI).

**Comment by ChrisBarker-NOAA** [comment]
> Is the goal to recommend a restricted set of units spellings and a restricted set of mathematical operators? If so, I believe it will require a units string grammar as well as the units database, both subsets of those found in UDUNITS.

I originally proposed two things:
1) document what is acceptable - this is essentially making a human readable form of the UDUNITS XML files (or Lars' UDSI).
 
2) have a set of recommended spellings, which would indeed be a "a units string grammar as well as the units database" as yu say.

But we could maybe keep it fairly simple and static (i.e. hand written) -- after all, "recommended" doesn't have to be complete -- as long as we cover most. 

And thinking now, we know what most units are -- that it, the ones that are canonical units in the standard names.

**Comment by ChrisBarker-NOAA** [comment]
Actually, looking now at UDSI again, I think we could create the rules pretty quickly, and then automate a full set -- at least of the ones in the standard names table.

Starting here as an example:


### metre m
UDUNITS2 alias names: meter
length
SIBaseUnit
**Definition:**  The metre, symbol m is the SI unit of length. It is defined by taking the fixed numerical value of the speed of light in vacuum, 𝑐, to be 299792458 when expressed in the unit m s−1 , where the second is defined in terms of the caesium frequency Δ𝜈Cs

### Recommendations:

These are the choices to make for recommendations:

 - US or Euro spelling? (meter or metre)
 - Spelled out or abbreviated? ("m" or "metre")
 - How to do math: 
     - Multiplication: UDUnits support a lot! 
        - Whitespace (space): newton meter
        - Period (.): newton.meter
        - Hyphen (-): newton-meter (maybe not actually allowed??) -- I haven't tested
        - Asterisk (*): newton*meter,
        - Centered Dot (·): newton·meter 
    - Exponentiation:
        - "hat" = m^2
        - "number" = m2
     - Division:
        - "Slash":  m/s
        - "negative exponent": m·s-1 

NOTE: I haven't tested all of those, and it's probably not a complete list, but you get the idea.

I think we could hash this out and produce something useful.

NOTE: I just found this page for what UDUNITS accepts:

https://ncics.org/portfolio/other-resources/udunits2/

I nice reference we could refer to, but doesn't really solve the problem at hand.

**Comment by larsbarring** [comment]
@ChrisBarker-NOAA: Thanks for moving this forward by making a concrete suggestion. I copy/paste you suggestion and add some comments:

>  - US or Euro spelling? (meter or metre)
> - Plural allowed? (meters) 
> - Spelled out or abbreviated? ("m" or "metre")

My suggestion is that we _do not standardarize_ this beyond what is needed for the CF Conventions texts and web pages etc.  (as per my previous [comment](https://github.com/cf-convention/cf-conventions/issues/562#issuecomment-2964358499)). What we however should standardise is the formatting of units and compound units when expressed with symbols. You outline many of the formatting variants accepted by UDUNITS2:


> - How to do math: 
>     - Multiplication: UDUnits support a lot! 
>        - Whitespace (space): newton meter
>        - Period (.): newton.meter
>        - Hyphen (-): newton-meter (maybe not actually allowed??) -- I haven't tested
>        - Asterisk (\*): newton\*meter,
>        - Centered Dot (·): newton·meter 
>    - Exponentiation:
>        - "hat" = m^2
>        - "number" = m2
>     - Division:
>        - "Slash":  m/s
>        - "negative exponent": m·s-1 

To my mild surprise `newon-meter` actually works for multiplication:
```bash
$ udunits2 -H newton-meter -W ""
    m²·kg·s⁻²
```
Moreover, UDUNITS2 allows `**` for exponentiation:
```bash
$ udunits2 -H m**2·kg·s**-2 -W "N m"
    1 m**2·kg·s**-2 = 1 (N m)
    x/(N m) = (x/m**2·kg·s**-2)
```
For division UDUNITS2 accepts multiple slashes, e.g.:
```bash
$ udunits2 -H "W m-2 sr-1 (m-1)-1" -W ""  # this is the canonical unit  toa_outgoing_radiance_per_unit_wavenumber*
    m·kg·s⁻³·rad⁻²
(base) [a001257@c21947 Downloads]$ udunits2 -H "W/m2/sr/(m)-1" -W ""
    m·kg·s⁻³·rad⁻²
(base) [a001257@c21947 Downloads]$ udunits2 -H "W/m2/sr/(m-1)" -W ""
    m·kg·s⁻³·rad⁻²
```
and rational, or real, numbers are not accepted as exponents.


I think that it is really useful that UDUNITS2 in itself is lenient towards the formatting (even though I personally do not like  the "newton-meter" variant). looking through the list of canonical units used the version 91 of the standard name table (see next comment) some guiding principles seems already in place:
- multiplication is represented by *space* (codepoint U+0020, decimal 32)
- exponentiation is represented by attaching the numeric exponent immediately after the unit symbol (what Chris calls a "number")
-  division is represented by negative numeric exponents (including "-1" for simple division), it is recommended not to use *slash* at all.
- parentheses ( *()* ) may be used when necessary.

I think that expressing these recommendations more clearly in the conventions text will serve us well. I addition to we should look through the list of canonical units currently in the standard name table (next comment).

**Comment by ChrisBarker-NOAA** [comment]
> I like the idea of recommending use of the unit symbols.

Me too -- we're converging! (well, three of us are, so far ....)

> However, I also think the recommendation should try to avoid non-ASCII characters where possible. 

I think it's always possible -- and a good idea.

Unicode is great, and things like an exponent are quite nice for human readability:  e.g. m²·kg·s⁻² is very nice to read -- much more so than m^2 kg s-2. But it's kind of a one way street -- there are many Unicode symbols that can mean the same thing (or similar things -- and that gets very confusing when you want computer-readable symbols.

So ASCII-only is a good idea in this case. (not because it's ASCII per-se, but because it means there's only one way to do it (or at least fewer ways to do it).

Hmm -- thinking now, an exception for a few non-ascii Symbols?

maybe: 
- The MIDDLE DOT U+00B7: ·
- The DEGREE SIGN (U+00B0): °
-  others??

However, I note that UNICODE also has: Degree Celsius (℃) and Degree Fahrenheit (℉) as single code points.

Which is another "more than one way to do it" -- so maybe ASCII only is the way to go for the recommendations.

And that would make the compliance checkers easier to write :-)

**Comment by larsbarring** [comment]
Elsewhere I have argued for not allowing anything but ASCII, but that was for attribute names, not the attribute value (content). While I do understand the argument against allowing unicode in unit symbols because poor software support. But the other side of this argument is -- from the software developer side -- that there is no need to support unicode because there is no dataset containing this, and even the [CF] standard does not allow it. I think that the only way to break this self-reinforcing circle is to start with the standard. It will take some time for both data producers and software software developers to take this on board, and CF should [shall] of course be aware of this transition time when making recommendations.

Regarding specific unicode characters:
1.  The DEGREE SIGN (U+00B0: °) is arguably the prime and most important candidate. By allowing this, and promoting it in the recommendations, we solve the unit symbol of [angular/phase] degree, degree Celsius  with just one unicode character. As a bonus the degree Fahrenheit and degree Rankin will be covered, but they do not appear in any canonical units.
2. As CF currently is using *space* for multiplication, we do not need the MIDDLE DOT (U+00B7: ·) in the recommendations. If someone is using it anyway then UDUNITS2 will handle it.
3. We should not be concerned with the unicode special symbols for degree Celsius (℃) and degree Fahrenheit (℉). SI uses the degree sign immediately followed by the latin uppercase letter C. Even more important is that the unicode standard itself explicitly discourage its use ([here](https://en.wikipedia.org/wiki/Celsius#Unicode_character)), the same goes for the Fahrenheit symbol ([here](https://en.wikipedia.org/wiki/Fahrenheit#Unicode_representation_of_symbol)).
4. We may want to consider the micro prefix: MICRO SIGN, U+00B5: µ), which is preferred by SI, and GREEK SMALL LETTER MU U+03BC: μ). However, this prefix is not part of any canonical unit. 
5. Other unicode unit symbols are:
     - Ohm: Greek Capital Letter Omega (U+03A9: Ω). There is a special ohm sign but it, too, is recommended against. Currently CF do not have an standard name requiring this.
      - Angular minute and angular second with symbols PRIME and DOUBLE PRIME. Currently CF do not have an standard name requiring this (the canonical unit is degree)
      - Ångström (1e-10 m) with symbol LATIN UPPERCASE LETTER A WITH RING ABOVE. There is a special Ångström sign but it, too, is recommended against. Currently CF do not have an standard name requiring this (the canonical unit is meter).
6. Finally, we have the superscript digits 0-9 and superscript plus and minus.

To sum up, if at all possible I think that we should seriously consider to recommend use of the DEGREE SIGN (U+00B0: °) in unit symbol strings. From the perspective of consistently using standard unit symbols in canonical units and all other unicode characters mare much less important.

**Comment by larsbarring** [comment]
@DocOtak I think we are focussing on different things here. 

To me, this discussion is _not_ (at all!) about at making a grammar for units strings. For that CF is stuck with what UDUNITS2 provides. However there are two practical problems with this:

1. For many units, UDUNITS2 allows one or several aliases names and/or symbols. Some of these are (nowadays) explicitly discouraged, e.g. by SI or by unicode (as detailed in my previous comment). Or they have never appeared outside UDUNITS2 (i.e. the climate/geophysical community). For this reason it is useful to provide recommendations to help users to choose the most standard and interoperable unit symbol.
2. Similarly, UDUNITS provide several alternative ways to combine simple units into compound ones (i.e. to "do unit maths" as Chris writes). As I wrote, in the standard name table table there are already well established "rules" that are not spelled out. These follow the rules provided by SI, as expressed in the SI Brochure, as do the alternative ways provided by UDUNITS2 as far as I reckon. 
   - My understanding of the CF rules for canonical units is as follows: _multiplication_: `space`; _exponentiation_: attach the numeric exponent (ASCII [0-9-+]) immediately after (no space) the unit symbol; _division_: multiplication with unit raised to  `-1` (do not use `/`); _parentheses_: allowed as necessary.
   -  Additionally UDUNITS allows: _multiplication_: `*` `.` `·` `-`; _exponentiation_:  `^`, `**` as well as unicode superscript roman digits and `⁺`(unicode superscript minus`⁻` currently hits a bug); _division_: one or several `/` 

I think that it is helpful to provide recommendations guiding users to avoid the more nonstandard (and expressly discouraged) variants available in 1, as well as the more ambiguous alternatives in 2 (e.g. multiple `/` without parentheses, or `-`for multiplication). 

In particular, I think that having in the standard name table canonical units like `degree s-1` that mix unit name with unit symbol is particularly problematic because this is expressly discouraged by SI. This is not something that CF should promulgate. Hence, my plea for allowing the degree sign (`°`).

**Comment by zdefne-usgs** [comment]
How does one type **°** on a standard keyboard? :) If **K** stands for Kelvin, why can we not use **C** for Celsius? I don't think there is any other unit that is denoted by **C**, right? 

Similarly, I would prefer typing in ohm rather than inserting a Unicode character. It's just easier.

**Comment by ChrisBarker-NOAA** [comment]
>  K stands for Kelvin, why can we not use C for Celsius? 

I was wondering about that -- in fact, I really thought the official spelling was simply "C" (and K, and F) -- but when I serached, it appears that SI does recommend the degree symbol.

> How does one type ° on a standard keyboard?

> I would prefer typing in ohm rather than inserting a Unicode character. It's just easier.

This is the non-ascii problem we've faced for years -- Apple has done a not too bad job of addressing it with the Emoji and Symbols tool -- but that only helps with GUI programs.

I frequently find myself copy+pasting it from somewhere ...

However, when generating a CF file -- how often are you hand-typing it? I'm sure most (all?) are generated by code, and you only have to write that code once, so not a huge burden.

And in Python, there are multiple options:

```
In [2]: "\u03A9, \N{Greek Capital Letter Omega}"
Out[2]: 'Ω, Ω'
```

**Comment by ChrisBarker-NOAA** [comment]
Thanks Lars. That settles it.

**Comment by ChrisBarker-NOAA** [comment]
@larsbarring wrote: "This appear to be a bug in UDUNITS2!"
- I've posted an issue on gitHub:
https://github.com/Unidata/UDUNITS-2/issues/128

Thought that doesn't look very active :-(

**Comment by ChrisBarker-NOAA** [comment]
@DocOtak:

> This thread was making me nervous

I don't think there's any reason for nerves here :-)

> CF should not get into the business of maintaining a quantities library.

Absolutely not -- that's what UDUNITS is, and why CF refers to it.

> If CF wants to make a grammar for unit strings, I think it needs to actually provide something like [ABNF]

Agreed, but we are not proposing creating a grammar.

>  I'm often frustrated by arguments claiming that something will be easy to program for without providing an actual program.

I, for one, sure know how hard it is !

The NCICS resource listing all the [contents of the udunits2 database](https://ncics.org/portfolio/other-resources/udunits2/) is great and I think should be provided by CF somehow as a resource to look at.

I agree, but it's not complete from the user's perspective. 

By that I mean it doesn't address how to combine units -- even for something as simple as a speed unit -- I can find meter there, and second, but not meter per second ...

(and to a lot of folks that haven't thought about it, they think of "m/s" as a unit -- even though it's actually a combination of units. e.g. :  https://github.com/Unidata/UDUNITS-2/issues/84)

> But most of the rest of the discussion seems aimed at someone who might be writing their own parser for the unit string, and my advice to someone thinking to do that is basically "don't", you'll probably waste less time trying to get udunits itself to run and interact with your language of choice. 

Indeed -- I have that exact problem and I'll probably punt to using udunits -- it's a lot easier now than it was. If I ever get there at all. But it still would be nice if there were *some* docs on what is accepted by UDUNITS -- and I don't think a proper grammer spec would be readable or helpful for most folks.

But no -- most of this discussion is the opposite:

What should we provide by way of advice to someone creating a new CF compliant file? How should they write their units out?

And I think it would be lot better for the community if we recommended a single way to do it.

I keep quoting Postels' law:

 "Be conservative in what you send, be liberal in what you accept"

UDUNITS is liberal in what it accepts, which is good, but CF should be conservative in what it sends, which it is not, if all we say is: "anything UDUNITS can read is fine" -- and that's all we say at this point.

**Comment by DocOtak** [comment]
@zdefne-usgs 
> How does one type ° on a standard keyboard?

If on a Mac, <kbd>option</kbd> + <kbd>shift</kbd> + <kbd>8</kbd>

This is from memory typing it... I didn't need to look it up. At least I'm "proud" to not remember any [alt codes](https://en.wikipedia.org/wiki/Alt_code).

---
Re Postels' Law:
I disagree with its applicability here in that CF does not _produce_ anything. CF might be said to _accept_ data in so far as a CF compliant file can be accepted by the standard. CF certainly doesn't send anything, data files are created by a massive cross disciplinary community

> What should we provide by way of advice to someone creating a new CF compliant file? How should they write their units out?
> And I think it would be lot better for the community if we recommended a single way to do it.

This still just sounds to me like developing either a strict grammar, or if you really want one way to do it, having a controlled vocabulary. NVS has [P06](https://vocab.nerc.ac.uk/collection/P06/current/) where the "alternate labels" are mostly udunits compatible. NVS also has [P24](https://vocab.nerc.ac.uk/collection/P24/current/) for dimensions. SDN has a URN for the units in a separate attribute and that's something I think is useful since it avoids the... "human mess" in the units string for parsing (I'm sorry I couldn't think of a better phrase here).

I don't think CF should, in the CF Document itself, provide guidance on what that unit string should be. We should link to that resource provided by NCICS, we could also recommend you check your unit strings against udunits and give the C meaning Coulomb not Celsius as an example of a potential foot gun. If the desire to have guidance is really going to be included, it needs to become more than guidance and become requirements.

---
@ChrisBarker-NOAA 
>I've posted an issue on gitHub:
>
>   https://github.com/Unidata/UDUNITS-2/issues/128
>
>Thought that doesn't look very active :-(

Indeed, if I were just examining this on github without the additional contact/communication with @ethanrd , the project would set off all the "dead" and "abandoned" project warning flags.

Udunits was unwilling or unable to address issues we've identified in the past, the temperature delta vs point on scale is what I have in mind here. Which does make me apprehensive about continuing to have it be the external thing CF points to which defines the acceptable unit strings. I do want to acknowledge here, that this situation is probably not the fault of any of the devs of udunits. Software takes time, energy, and smarts to maintain, and its often for a thankless userbase.

I wonder now... how much of my apprehension about this entire discussion is driven by concern about the future of udunits itself. At the same time, the data management circles I'm involved with are becoming increasingly intolerant of free form fields that are meant for computer interpretation. Everything seems to be moving toward controlled vocabularies.

**Comment by zdefne-usgs** [comment]
> Just a quick reply:
> 
> $ udunits2 -H C -W ""
>     s·A
> $ udunits2 -H F -W ""
>     m⁻²·kg⁻¹·s⁴·A²
> That is, `C` is already the symbol for coulomb that is the unit for electrical charge, and similarly `F` is already the symbol for farad that is the unit for electrical capacitance.
> 
> However, with the degree symbol both works as expected:
> 
> $ udunits2 -H °C  -W ""
>     K @ 273.15
> $ udunits2 -H °F -W ""
>     0.555555555555556 K @ 459.67

That settles it, need a "degree" in there. Though, isn't it peculiar that it is °C but °K? Interestingly, if you go to the origins of it, **it was indeed °K** . 
https://www.bipm.org/en/-/resolution-cgpm-11-12

So, I guess the point is that people are inclined to go with the easier option, and somewhere along the lines the degree sign was dropped from °K. Therefore, it is important for the conventions or standards to consider usability as well. Hence, I believe degree_C is more preferable than °C because of its ease of use.

**Comment by ChrisBarker-NOAA** [comment]
I'm not sure ASCII only is the way to go, but the ability to do everything with jsut ASCII probably is, so while I'd love "only one way to do it":

Either degreeC  or ºC -- but nothing else is recommended.

Except:

```
You have: 1 ºC
udunits2: Don't recognize "ºC"
```
But it does recognize the combined symbol :-( 

```
You have: 1 ℃
You want: 
    K @ 273.15
```

Darn!

**Comment by larsbarring** [comment]
Chris, @ChrisBarker-NOAA, the degree sign (unicode u+00B0) works just fine for me:  

```
$ udunits2
You have: 1 °C  
You want: K
    1 °C = 274.15 K
    x/K = (x/°C) + 273.15
You have: 2 °F
You want: °C
    2 °F = -16.6667 °C
    x/°C = 0.555556*(x/°F) - 17.7778
```

So I am not sure what is going on, different versions ???

And, to your second post, yes this seems inconsistent, which is perfectly acceptable by UDUNITS2 as input (which is fine), but I daresay not something that CF -- or anyone else for that matter -- should promote for producing data. In particular, the units symbols that the unicode themselves discourage (e.g. https://en.wikipedia.org/wiki/Celsius#Unicode_character)

**Comment by DocOtak** [comment]
The netCDF4-python library will force attributes containing unicode points outside the ASCII range to be of type STRING in netCDF4 files. This has been the case for almost a decade now. To me, this basically makes the suggestion to use ° anywhere in a CF defined attribute impossible to follow and remain valid CF.

**Comment by ethanrd** [comment]
Quick side-note regarding spelling, I just wanted to point out these two statements.
1) [CF Section 2.3](https://cfconventions.org/cf-conventions/cf-conventions.html#_naming_conventions) "Naming Convenions" does say
   > Attribute names and their contents, where standardized, are given in English in this document and should appear in English in conforming netCDF files for the sake of portability.

2) And the CF Standard Names Guidelines document [section on spelling](https://cfconventions.org/Data/cf-standard-names/docs/guidelines.html#id2797709) (which isn't really applicable for units strings but figured I'd point it out) says:
   > US spelling is used e.g. vapor, sulfur.

**Comment by ChrisBarker-NOAA** [comment]
> Chris, @ChrisBarker-NOAA, the degree sign (unicode u+00B0) works just fine for me:

Yup -- my mistake -- wow! this is a pain -- turns out that character is not the degree symbol, but rathter:
'MASCULINE ORDINAL INDICATOR'

I got that from the Mac "Emoji and Symbols" Character Viewer -- I must. have made the mistake once, and then it show up in the "frequently used" list.

Time to be more careful!

**Comment by ethanrd** [comment]
From [Section 2.2](https://cfconventions.org/cf-conventions/cf-conventions.html#_data_types) "Data Types":
> A text string can be stored either in a variable-length string or in a fixed-length char array. In both cases, text strings must be represented in Unicode Normalization Form C (NFC, section 3.11 and Annex 15 of the Unicode standard) and encoded according to UTF-8.

That section seems pretty focused on data variables. However, there is a mention of attributes is regarding n-dimensional arrays of strings:
> An n-dimensional array of strings may be implemented as a variable or an attribute of type string with n dimensions (only n=1 is allowed for an attribute) or as a variable of type char with n+1 dimensions ...

My interpretation is that attributes can be 'string' or 'char' and must be encode as UTF-8. However, it would be nice if attributes were more explicitly discussed here (in general, not just one edge case).

**Comment by DocOtak** [comment]
I forgot that attributes of type string were added: https://github.com/cf-convention/cf-conventions/pull/556

**Comment by larsbarring** [comment]
Could we find a moderator to help summarize the salient points and arguments of this conversation?

**Comment by larsbarring** [comment]
There are a lot of things going on in the units [of measure] world right now. Hence, I think this issue should not be closed as dormant. 

By the way, UDUNITS (source code main branch) was recently updated to properly parse unicode exponents, handling of inf/nan  (as in infinity/not-a-number) and nano (prefix), and "per" (indicating ratio). 

Moreover, the UDUSI website was recently updated with search functionality (and total rewrite of the underlying machinery). Now it now lives here: **https://udusi.github.io/** (the old url will redirect).

**Comment by ChrisBarker-NOAA** [comment]
> Moreover, the UDUSI website was recently updated with search functionality (and total rewrite of the underlying machinery). Now it now lives here: https://udusi.github.io/

Great! maybe the first step is to provide a link to that site in the docs. 

I'd love it to be a bit more easy and obvious that UDUSI is a massive improvement over the "just see if it works with UDUNITS" approach!
